### PR TITLE
wip: fix: Get Velero Azure plugin operational

### LINKED ISSUE #7450: Velero fails to start for AKS boot due to misconfigured BackupStorageLocation
### Summary 

Currently working through getting Velero operational on AKS and some `jx-requirements.yml` config needs enhancing. ResourceGroup and BucketName (rather than URL) are required for the Velero Azure Plugin.

### Steps to reproduce the behavior

`jx boot` with Velero enabled on an AKS cluster

### Expected behavior

Velero pod doesn't crash out

### Actual behavior

Velero pod crashes out because of mandatory resourceGroup config and using bucketUrl rather than bucketName

### Jx version

The output of `jx version` is:

```
COPY OUTPUT HERE
```

### Diagnostic information

The output of `jx diagnose version` is:

```
COPY OUTPUT HERE
```

### Kubectl version
The output of `kubectl version --client` is:

```
Copy output here
```

### PR DESCRIPTION
**Pull request by chrismellard**
#### Submitter checklist

- [X] Change is code complete and matches issue description.
- [X] Change is covered by existing or new tests.

#### Description

Adds in new requirementsConfig keys to support Azure Velero plugin. Further changes required to be made to boot repo. Need to implement part of the BucketProvider interface to get through the boot process. Only those methods required to get through `jx boot` implemented for now

#### Which issue this PR fixes

fixes #7450

### CODE REVIEW

**Comment by chrismellard** [comment]
Related jenkins-x-boot-config PR here jenkins-x/jenkins-x-boot-config#168

**Comment by ankitm123** [comment]
I know the codebase does not have good test coverage, but would be nice if you can add some tests for the code u wrote.

**Comment by ankitm123** [comment]
Also, you should squash your commits, and run `make lint`, I see some goimport issues :)

File: pkg/cloud/aks/storage/bucket_provider.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+import (
+	"fmt"
+	"github.com/jenkins-x/jx-logging/pkg/log"
+	"github.com/jenkins-x/jx/v2/pkg/cloud/aks"
+	"github.com/jenkins-x/jx/v2/pkg/cloud/buckets"
+	"github.com/jenkins-x/jx/v2/pkg/config"
+	"github.com/jenkins-x/jx/v2/pkg/util"
+	"github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
```
[the remark this answers is not in the document]
Comment (ankitm123) [reply pkg/cloud/aks/storage/bucket_provider.go:11]: Created a PR to fix it, moving forward we plan to use the google uuid package: https://github.com/google/uuid

**Comment by chrismellard** [comment]
Thanks. Left the two commits in there as they were addressing different concerns - kinda. But happy to squash them up with addressing your feedback. Try get something turned around tonight.

File: pkg/cloud/aks/storage/bucket_provider.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
+package storage
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	uuid "github.com/google/uuid"
```
Comment (ankitm123) [inline pkg/cloud/aks/storage/bucket_provider.go:8]: Dont need the uuid alias.

File: pkg/cloud/aks/storage/bucket_provider_test.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
+// +build unit
```
Comment (ankitm123) [inline pkg/cloud/aks/storage/bucket_provider_test.go:1]: :100:

File: pkg/cloud/aks/mocks/azurestorage.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
+// Code generated by pegomock. DO NOT EDIT.
+// Source: github.com/jenkins-x/jx/v2/pkg/cloud/aks (interfaces: AzureStorage)
+
+package aks_test
+
+import (
```
Comment (ankitm123) [inline pkg/cloud/aks/mocks/azurestorage.go:6]: So, this is just one of the side-effects of using pegomock, you need to fix the goimport issue (sadly in a file that should not be edited by hand), for the codegen tests to pass. I have an [issue](https://github.com/jenkins-x/jx/issues/7307) to replace pegomock with gomock.

**Comment by ankitm123** [comment]
I think you should rebase this PR with the latest master, there was a PR which fixed issues with builds: https://github.com/jenkins-x/jx/pull/7460

**Comment by chrismellard** [comment]
🤷

**Comment by rawlingsj** [comment]
this is the unit test failure:
```
--- FAIL: TestNoForkAndExistingDir (0.20s)
    helpers_test.go:1137: 
        	Error Trace:	helpers_test.go:1137
        	            				helpers_test.go:385
        	Error:      	Received unexpected error:
        	            	failed to run 'git cherry-pick 9492c14c9b08ddd9f2b1230fc486fb154955bd48' command in directory '/tmp/177040746', output: 'The previous cherry-pick is now empty, possibly due to conflict resolution.
        	            	If you wish to commit it anyway, use:
        	            	
        	            	    git commit --allow-empty
        	            	
        	            	Otherwise, please use 'git reset'
        	            	On branch master
        	            	You are currently cherry-picking commit 9492c14.
        	            	
        	            	nothing to commit, working tree clean'
        	            	git output: The previous cherry-pick is now empty, possibly due to conflict resolution.
        	            	If you wish to commit it anyway, use:
        	            	
        	            	    git commit --allow-empty
        	            	
        	            	Otherwise, please use 'git reset'
        	            	On branch master
        	            	You are currently cherry-picking commit 9492c14.
        	            	
        	            	nothing to commit, working tree clean
        	            	github.com/jenkins-x/**/v2/pkg/gits.(*GitCLI).gitCmd
        	            		/workspace/source/pkg/gits/git_cli.go:598
        	            	github.com/jenkins-x/**/v2/pkg/gits.(*GitCLI).CherryPick
        	            		/workspace/source/pkg/gits/git_cli.go:1246
        	            	github.com/jenkins-x/**/v2/pkg/gits.ForkAndPullRepo
        	            		/workspace/source/pkg/gits/helpers.go:681
        	            	github.com/jenkins-x/**/v2/pkg/gits_test.runForkAndPullTestCase
        	            		/workspace/source/pkg/gits/helpers_test.go:1130
        	            	github.com/jenkins-x/**/v2/pkg/gits_test.TestNoForkAndExistingDir
        	            		/workspace/source/pkg/gits/helpers_test.go:385
        	            	testing.tRunner
        	            		/usr/local/go/src/testing/testing.go:909
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1357
        	            	Unable to cherry-pick 9492c14c9b08ddd9f2b1230fc486fb154955bd48
        	            	github.com/jenkins-x/**/v2/pkg/gits.ForkAndPullRepo
        	            		/workspace/source/pkg/gits/helpers.go:690
        	            	github.com/jenkins-x/**/v2/pkg/gits_test.runForkAndPullTestCase
        	            		/workspace/source/pkg/gits/helpers_test.go:1130
        	            	github.com/jenkins-x/**/v2/pkg/gits_test.TestNoForkAndExistingDir
        	            		/workspace/source/pkg/gits/helpers_test.go:385
        	            	testing.tRunner
        	            		/usr/local/go/src/testing/testing.go:909
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1357
        	Test:       	TestNoForkAndExistingDir
    helpers_test.go:1141: 
        	Error Trace:	helpers_test.go:1141
        	            				helpers_test.go:385
        	Error:      	Not equal: 
        	            	expected: "/tmp/177040746"
        	            	actual  : ""
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-/tmp/177040746
        	            	+
        	Test:       	TestNoForkAndExistingDir
    helpers_test.go:1142: 
        	            	actual  : ""
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-master
        	            	+
        	Test:       	TestNoForkAndExistingDir
    helpers_test.go:1143: 
        	Error Trace:	helpers_test.go:1143
        	            				helpers_test.go:385
        	Error:      	Not equal: 
        	            	expected: &gits.GitRepository{ID:0, Name:"roadrunner", AllowMergeCommit:false, HTMLURL:"https://fake.git/acme/roadrunner", CloneURL:"file:///tmp/660574327/acme", SSHURL:"", Language:"", Fork:false, Stars:0, URL:"https://fake.git/acme/roadrunner.git", Scheme:"https", Host:"fake.git", Organisation:"acme", Project:"", Private:false, HasIssues:false, OpenIssueCount:0, HasWiki:false, HasProjects:false, Archived:false}
        	            	actual  : (*gits.GitRepository)(nil)
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,23 +1,2 @@
        	            	-(*gits.GitRepository)({
        	            	- ID: (int64) 0,
        	            	- Name: (string) (len=10) "roadrunner",
        	            	- AllowMergeCommit: (bool) false,
        	            	- HTMLURL: (string) (len=32) "https://fake.git/acme/roadrunner",
        	            	- CloneURL: (string) (len=26) "file:///tmp/660574327/acme",
        	            	- SSHURL: (string) "",
        	            	- Language: (string) "",
        	            	- Fork: (bool) false,
        	            	- Stars: (int) 0,
        	            	- URL: (string) (len=36) "https://fake.git/acme/roadrunner.git",
        	            	- Scheme: (string) (len=5) "https",
        	            	- Host: (string) (len=8) "fake.git",
        	            	- Organisation: (string) (len=4) "acme",
        	            	- Project: (string) "",
        	            	- Private: (bool) false,
        	            	- HasIssues: (bool) false,
        	            	- OpenIssueCount: (int) 0,
        	            	- HasWiki: (bool) false,
        	            	- HasProjects: (bool) false,
        	            	- Archived: (bool) false
        	            	-})
        	            	+(*gits.GitRepository)(<nil>)
        	            	 
        	Test:       	TestNoForkAndExistingDir
    helpers_test.go:1149: 
        	Error Trace:	helpers_test.go:1149
        	            				helpers_test.go:385
        	Error:      	"[]" should have 1 item(s), but has 0
        	Test:       	TestNoForkAndExistingDir
FAIL

```

**Comment by rawlingsj** [comment]
The boot and terraform BDD tests are failing to deploy a quickstart into the staging environment, the staging master pipeline has the following error:
```
error: error getting Requirements from TeamSettings to determine if in GitHub app mode: validation failures in requirements from team settings:
velero: Additional property bucketName is not allowed
velero: Additional property resourceGroup is not allowed

```

**Comment by rawlingsj** [comment]
@chrismellard looking at the PR is it fair to say it provides a way to create buckets in azure storage?  I wonder if this should be done as part of Terraform instead.  

There has been an effort of work this year to decouple and move the creation + management of cloud resources out of the jx CLI and into Terraform.  I wonder if that would be a better approach?  What would be the main challenges doing that?

 In the future we'd like to support other means of creating clusters too, https://crossplane.io/ looks really interesting as well but the more we can decouple cloud provider related commands from core jx the easier things like that will be.

WDYT?

**Comment by chrismellard** [comment]
Unit test appears to have been failing for a while and I'm fixing up in another PR, seems to be related to git error messages based on empty commits  for rebases vs cherrypicks... though surprised given this PR contains that fix this is still failing unit tests (it certainly rectifies *a* unit test failure locally)

Happy to move these cloud resource creations in to terraform. Was unaware of the initiative to move this logic in to terraform ;) I'll have a think and revise this PR (+ terraform module)

**Comment by rawlingsj** [comment]
Ok sounds good, we need to do a better job of communicating these initiatives, I'll add an agenda item for the office hours next week to discuss ways we need to improve.

**Comment by chrismellard** [comment]
Unit test PR here btw - #7468

**Comment by chrismellard** [comment]
Having thought about this and with a view to not break other public cloud implementations how about we create a structure of cloud capabilities per cloud cluster, i.e. what we are or are not leveraging from terraform modules. Can start to data drive the code from this rather than hard coding `!= GKE` everywhere. For this PR this will initially just contain logic informing the verify step to not create the Azure storage URLs defined in `jx-requirements` but just to verify their existence. We'll have a structure somewhere reflecting the fact that `AKS` has its storage provisioned via terraform.

Does that sound like an approach everyone is happy with?

**Comment by chrismellard** [comment]
Will reopen another PR for this when it's nearer completion

**Pull request closed without merging**